util/src/main/java/com/google/cloud/hadoop/util/interceptors/InvocationIdInterceptor.java

```diff
@@ -0,0 +1,51 @@
+package com.google.cloud.hadoop.util.interceptors;
+
+import com.google.api.client.http.HttpExecuteInterceptor;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpRequest;
+import java.io.IOException;
+import java.util.UUID;
+
+public final class InvocationIdInterceptor implements HttpExecuteInterceptor {
+
+  public static final String GCCL_INVOCATION_ID_PREFIX = "gccl-invocation-id/";
+
+  public static final String GOOG_API_CLIENT = "x-goog-api-client";
+
+  private final HttpExecuteInterceptor interceptor;
+
+  public InvocationIdInterceptor(HttpExecuteInterceptor interceptor) {
+    this.interceptor = interceptor;
+  }
+
+  @Override
+  public void intercept(HttpRequest request) throws IOException {
+    if (this.interceptor != null) {
+      this.interceptor.intercept(request);
+    }
+    HttpHeaders headers = request.getHeaders();
+    String existing = (String) headers.get(GOOG_API_CLIENT);
+    if (isInvocationIdPresent(existing)) {
+      // As invocationId is already present,It's a retried request.
+      // TODO: add support for attempt_count
+      return;
+    }
+    UUID invocationId = UUID.randomUUID();
```

Comment:
did we consider having this based on timestamp, rather than having this to be fully random? i.e. using the invocation id, we can get some idea about then the call was made for eg.

Reply:
This is not the requirement. GCS do get timestamp information attached to every request already.